### ISSUE TITLE
NullReferenceException

### ISSUE THREAD

**Issue by ozrecsec** [jchristn/WatsonTcp#129]
Hi, sometimes in my application i get this Exception from WatsonTCP. 

```
A Task's exception(s) were not observed either by Waiting on the Task or accessing its Exception property. As a result, the unobserved exception was rethrown by the finalizer thread.\n System.NullReferenceException: Object reference not set to an instance of an object.
   at WatsonTcp.WatsonTcpClient.<DataReceiver>d__62.MoveNext() \n \n\nException: System.AggregateException: A Task's exception(s) were not observed either by Waiting on the Task or accessing its Exception property. As a result, the unobserved exception was rethrown by the finalizer thread. ---> System.NullReferenceException: Object reference not set to an instance of an object.
   at WatsonTcp.WatsonTcpClient.<DataReceiver>d__62.MoveNext()
   --- End of inner exception stack trace ---
---> (Inner Exception #0) System.NullReferenceException: Object reference not set to an instance of an object.
   at WatsonTcp.WatsonTcpClient.<DataReceiver>d__62.MoveNext()<---
```

My aplpication is a WPF solution and I'm catching all uncatched exceptions in App.xaml.cs with 

```
AppDomain.CurrentDomain.UnhandledException += CurrentDomain_UnhandledException;
Dispatcher.UnhandledException += Dispatcher_UnhandledException;
TaskScheduler.UnobservedTaskException += TaskScheduler_UnobservedTaskException;
Application.Current.DispatcherUnhandledException += Current_DispatcherUnhandledException;
```

This apparently happens "randomly" , i think it's more frequent when I close the connection.

Have you any idea about this?
thanks a lot for your great work

**Comment by ozrecsec** [comment]
I've noticed that the exception occurs in my client when I stop the server and then I reopen it. The server is a WatsonTCPServer.
At this time my client reconnects automatically to the server and initially it receives this Exception (It tries to reconnect every 3 seconds).
It's strange but seems like the client has some old incoming data from the previous connection, but it's strange because when the client receives  the ServerDisconnected event it will be disposed and to try reconnection it will instantiate new WatsonTcpClient.
My client is a WPF application running on Windows 2012R2, .NET framework 4.6.1
My server is a console application with TopShelf and running on same machine and same framework

**Comment by jchristn** [comment]
Hi @ozrecsec thanks for your note.  This one is very interesting; the entire ```DataReceiver``` on the client is wrapped in ```try/catch``` and exceptions are never re-thrown.  

If I were to **guess**, I would say it's probably this code in the ```finally``` block:
```
_Settings.Logger?.Invoke(Severity.Error,
    _Header + "data receiver exception for " + _ServerIp + ":" + _ServerPort + ":" + Environment.NewLine + SerializationHelper.SerializeJson(e, true) + Environment.NewLine);

_Events.HandleExceptionEncountered(this, new ExceptionEventArgs(e));
break;
```

When the client is disposed, this happens:
```
_Settings = null;
_Events = null;
```

Which would lead to a null reference exception.

Would you have a chance to JSON-serialize the entirety of the exception and paste it here?  

Alternatively, you could try encapsulating the statements in the ```DataReceiver``` ```catch``` block like this to see if it resolves the problem?  If it does, I can put together a patch straight away (I would only ask you do to this because I don't have your environment):
```
catch (Exception e)
{
    if (_Settings != null)
    	_Settings.Logger?.Invoke(Severity.Error,
        	_Header + "data receiver exception for " + _ServerIp + ":" + _ServerPort + ":" + Environment.NewLine + SerializationHelper.SerializeJson(e, true) + Environment.NewLine);

    if (_Events != null)
    	_Events.HandleExceptionEncountered(this, new ExceptionEventArgs(e));
    	
    break;
} 
```

Cheers, Joel

**Comment by jchristn** [comment]
Actually, I'll do this now and post a patch.  Won't take long, give me a moment.  It needs to be fixed anyway :)

**Comment by ozrecsec** [comment]
I've tried the new version but the exception is still here :-( 

below the exception serialized as JSON

I've noticed that te two last lines of DataReceiver method are using _Settings and _Events and they are not in a try/catch block

```
{
  "ClassName": "System.AggregateException",
  "Message": "A Task's exception(s) were not observed either by Waiting on the Task or accessing its Exception property. As a result, the unobserved exception was rethrown by the finalizer thread.",
  "Data": null,
  "InnerException": {
    "ClassName": "System.NullReferenceException",
    "Message": "Object reference not set to an instance of an object.",
    "Data": null,
    "InnerException": null,
    "HelpURL": null,
    "StackTraceString": "   at WatsonTcp.WatsonTcpClient.<DataReceiver>d__62.MoveNext()",
    "RemoteStackTraceString": null,
    "RemoteStackIndex": 0,
    "ExceptionMethod": "8\nMoveNext\nWatsonTcp, Version=4.8.3.0, Culture=neutral, PublicKeyToken=null\nWatsonTcp.WatsonTcpClient+<DataReceiver>d__62\nVoid MoveNext()",
    "HResult": -2147467261,
    "Source": "WatsonTcp",
    "WatsonBuckets": null
  },
  "HelpURL": null,
  "StackTraceString": null,
  "RemoteStackTraceString": null,
  "RemoteStackIndex": 0,
  "ExceptionMethod": null,
  "HResult": -2146233088,
  "Source": null,
  "WatsonBuckets": null,
  "InnerExceptions": [
    {
      "ClassName": "System.NullReferenceException",
      "Message": "Object reference not set to an instance of an object.",
      "Data": null,
      "InnerException": null,
      "HelpURL": null,
      "StackTraceString": "   at WatsonTcp.WatsonTcpClient.<DataReceiver>d__62.MoveNext()",
      "RemoteStackTraceString": null,
      "RemoteStackIndex": 0,
      "ExceptionMethod": "8\nMoveNext\nWatsonTcp, Version=4.8.3.0, Culture=neutral, PublicKeyToken=null\nWatsonTcp.WatsonTcpClient+<DataReceiver>d__62\nVoid MoveNext()",
      "HResult": -2147467261,
      "Source": "WatsonTcp",
      "WatsonBuckets": null
    }
  ]
}
```

**Comment by ozrecsec** [comment]
Sorry, at moment I have problems with GitHub's SSL certificate, maybe I've to talk with my network administrator. But I'ts not so quick to do :-) .
So I must wait your next release to check if it works.

**Comment by ozrecsec** [comment]
See the pullrequest #130

**Comment by ozrecsec** [comment]
Finally I found the real problem. in my code on client side I call 
```
Client.Disconnect();
Client.Dispose();
```
If I put "Task.Delay(1000).Wait();" between Disconnect() and Dispose() it works fine.

I've seen that I can call only  .Dispose() and then the .Disconnect() is called internally and so it produces the same problem.

Then I tried to put at the end of Disconnect method in WatsonTcpClient a wait to the end of _DataReceiver task before setting the "Connected" variable to false. This way it works.
I'll do a pullrequest with my changes.
